src/Type.h

```diff
@@ -90,25 +93,25 @@ struct halide_handle_cplusplus_type {
 
     /// Qualifiers and indirections on type. Allows up to 8 levels. 0 is innermost.
     struct CPPTypeModifiers {
-        uint8_t data[8];
-      CPPTypeModifiers(const std::vector<uint8_t> &vals) {
-            user_assert(vals.size() <= sizeof(data)) << "Too many levels of indirection in handle type " << vals.size() << " where " << sizeof(data) << " are allowed\n";
-            std::copy(vals.begin(), vals.end(), &data[0]);
-            std::fill(&data[vals.size()], &data[sizeof(data)], 0);
+        std::array<uint8_t, 8> data;
+        CPPTypeModifiers(const std::vector<uint8_t> &vals) {
+            user_assert(vals.size() <= data.size()) << "Too many levels of indirection in handle type " << vals.size() << " where " << sizeof(data) << " are allowed\n";
+            std::copy(vals.begin(), vals.end(), data.begin());
+            std::fill(data.begin() + vals.size(), data.end(), 0);
         }
         const uint8_t &operator[](size_t index) const { return data[index]; }
         bool operator==(const CPPTypeModifiers &rhs) const {
-            return std::equal(&data[0], &data[sizeof(data)], &rhs.data[0]);
+            return std::equal(data.begin(), data.end(), rhs.data.begin());
         }
         const uint8_t *begin() const { return &data[0]; }
-        const uint8_t *end() const { return &data[sizeof(data)]; }
+        const uint8_t *end() const { return begin() + data.size(); }
```

Comment:
I'm giving up on msvc 2013. Trunk llvm no longer builds with it.

On Fri, Jul 29, 2016 at 11:17 AM, Zalman Stern <notifications@github.com>
wrote:

> In src/Type.h
> <https://github.com/halide/Halide/pull/1413#discussion_r72837503>:
>
> >          }
> >          const uint8_t *begin() const { return &data[0]; }
> > -        const uint8_t *end() const { return &data[sizeof(data)]; }
> > +        const uint8_t *end() const { return begin() + data.size(); }
>
> Please change this to "&data[0] + data.size()" to make it clear what is
> going on.
>
> IIRC, the general reason I bailed on std::array is because it is
> impossible to construct the end pointer while staying inside the std::array
> interface. I.e. all this code here placates Microsoft's compiler warnings
> while offering no actual protection because one has to convert to pointers
> anyway.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/halide/Halide/pull/1413/files/f44d10d08906741c216cb1246d0db68c587639c0#r72837503>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfdRsBrYidyqnrRTFEBKymfH1oXZRZxks5qakOsgaJpZM4JXvRH>
> .
>